File: war/js/d3/plotly-histogram-chart.js
/*
 *  Copyright 2019 Information and Computational Sciences,
 *  The James Hutton Institute.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

function plotlyHistogramChart() {
	var onPointsSelected = null,
		height = 600,
		xAxisTitle = '',
		yAxisTitle = '',
		colors = ["#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f", "#bcbd22", "#17becf"];

	function chart(selection) {
		selection.each(function (rows) {
			var data = [];
			var layout = {
				height: height,
				margin: {autoexpand: true},
				selectdirection: 'h',
				dragmode: 'select',
				legend: {
					orientation: 'h'
				},
				xaxis: {
					title: xAxisTitle
				},
				yaxis: {
					title: yAxisTitle
				}
			};

			var dims = Object.keys(rows[0]);

			dims.forEach(function (c, i) {
				data.push({
					x: unpack(rows, c),
					type: 'histogram',
					name: c,
					marker: {
						color: colors[i % colors.length]
					}
				});
			});

			var config = {
				modeBarButtonsToRemove: ['toImage'],
				displayModeBar: true,
				responsive: true,
				displaylogo: false
			};

			Plotly.newPlot(this, data, layout, config);

			var that = this;

			this.on('plotly_selected', function (eventData) {
				if (!eventData || (eventData.points.length < 1)) {
					Plotly.restyle(that, {selectedpoints: null});
				} else {
					if (onPointsSelected)
						onPointsSelected(eventData.points);
				}
			});
		});
	}

	function unpack(rows, key) {
		return rows.map(function (row) {
			return row[key];
		});
	}

	chart.height = function (_) {
		if (!arguments.length) return height;
		height = _;
		return chart;
	};

	chart.colors = function (_) {
		if (!arguments.length) return colors;
		colors = _;
		return chart;
	};

	chart.xAxisTitle = function (_) {
		if (!arguments.length) return xAxisTitle;
		xAxisTitle = _;
		return chart;
	};

	chart.yAxisTitle = function (_) {
		if (!arguments.length) return yAxisTitle;
		yAxisTitle = _;
		return chart;
	};

	chart.onPointsSelected = function (_) {
		if (!arguments.length) return onPointsSelected;
		onPointsSelected = _;
		return chart;
	};

	return chart;
}

File: src/jhi/germinate/client/widget/d3js/PlotlyHistogramChart.java
/*
 *  Copyright 2019 Information and Computational Sciences,
 *  The James Hutton Institute.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package jhi.germinate.client.widget.d3js;

import com.google.gwt.user.client.ui.*;

import jhi.germinate.client.page.*;
import jhi.germinate.shared.*;

/**
 * @author Sebastian Raubach
 */
public class PlotlyHistogramChart extends AbstractChart implements PlotlyChart
{
	private boolean needsRedraw = true;
	private String xAxisTitle = "";
	private String yAxisTitle = "";

	@Override
	protected void createContent(FlowPanel chartPanel)
	{
		panel.add(chartPanel);

		if (!StringUtils.isEmpty(filePath))
			onResize(true, false);
	}

	@Override
	public int[] getDownloadSize()
	{
		return new int[]{1280, 600};
	}

	@Override
	public void onResize(boolean containerResize, boolean force)
	{
		if (force || (needsRedraw && !StringUtils.isEmpty(filePath)))
		{
			needsRedraw = false;
			super.onResize(containerResize, force);
		}
	}

	@Override
	protected void updateChart(int width)
	{
		create(width);
	}

	@Override
	protected String getPhotoExportFilename()
	{
		return "histogram";
	}

	@Override
	protected MenuItem[] getAdditionalMenuItems()
	{
		return null;
	}

	@Override
	public Library[] getLibraries()
	{
		return new Library[]{Library.PLOTLY, Library.PLOTLY_HISTOGRAM_CHART, Library.D3_DOWNLOAD};
	}

	public String getxAxisTitle()
	{
		return xAxisTitle;
	}

	public PlotlyHistogramChart setxAxisTitle(String xAxisTitle)
	{
		this.xAxisTitle = xAxisTitle;
		return this;
	}

	public String getyAxisTitle()
	{
		return yAxisTitle;
	}

	public PlotlyHistogramChart setyAxisTitle(String yAxisTitle)
	{
		this.yAxisTitle = yAxisTitle;
		return this;
	}

	private native void create(int widthHint)/*-{
		var filePath = this.@jhi.germinate.client.widget.d3js.AbstractChart::filePath;
		var panelId = this.@jhi.germinate.client.widget.d3js.AbstractChart::panelId;
		var height = Math.round(@jhi.germinate.client.util.JavaScript.D3::HEIGHT * 0.8);
		var colors = @jhi.germinate.client.util.JavaScript.D3::getColorPalette()();
		var xAxisTitle = this.@jhi.germinate.client.widget.d3js.PlotlyHistogramChart::xAxisTitle;
		var yAxisTitle = this.@jhi.germinate.client.widget.d3js.PlotlyHistogramChart::yAxisTitle;

		$wnd.d3.tsv(filePath, function (data) {
			$wnd.d3.select("#" + panelId)
				.datum(data)
				.call($wnd.plotlyHistogramChart()
					.xAxisTitle(xAxisTitle)
					.yAxisTitle(yAxisTitle)
					.height(height)
					.colors(colors));
		});
	}-*/;
}
